File: csharp/2015/day02a/Program.cs
using System;
using System.IO;

class Program
{
    static void Usage()
    {
        Console.Error.WriteLine("usage: dotnet run <input file>");
        Environment.Exit(1);
    }

    static int ParseValue(string s)
    {
        if ( !int.TryParse(s, out int value) )
        {
            Console.Error.WriteLine("Error parsing string as number: " + s);
            Environment.Exit(1);
        }
        return value;
    }

    static int Process(string filename)
    {
        int totalArea = 0;

        try
        {
            using ( StreamReader reader = new StreamReader(filename) )
            {
                string? line;
                while ( (line = reader.ReadLine()) != null )
                {
                    string[] parts = line.Split('x');
                    int l = ParseValue(parts[0]);
                    int w = ParseValue(parts[1]);
                    int h = ParseValue(parts[2]);
                    int area1 = l * w;
                    int area2 = l * h;
                    int area3 = w * h;
                    int surfaceArea = 2 * area1 + 2 * area2 + 2 * area3;
                    int minArea = Math.Min(area1, Math.Min(area2, area3));
                    totalArea += surfaceArea + minArea;
                }
            }
        }
        catch ( IOException e )
        {
            Console.Error.WriteLine("Error reading file: " + e.Message);
            Environment.Exit(1);
        }

        return totalArea;
    }

    static void Main(string[] args)
    {
        if ( args.Length < 1 )
        {
            Usage();
        }

        string filename = args[0];
        int result = Process(filename);
        Console.WriteLine("result = " + result);
    }
}

File: fsharp/2015/day02a/Program.fs
open System
open System.IO

let usage() =
    Console.Error.WriteLine("usage: dotnet run <input file>")
    Environment.Exit(1)

let processFile(filename: string) =
    let f acc (line: string) =
        let dims = line.Split('x') |> Array.map int
        match dims with
        | [| l; w; h |] ->
            let area1 = l * w
            let area2 = l * h
            let area3 = w * h
            let surfaceArea = 2 * area1 + 2 * area2 + 2 * area3
            let minArea = List.min [area1; area2; area3]
            acc + surfaceArea + minArea
        | _ ->
            failwithf "unexpected input: %s" line

    try
        let content = File.ReadAllText(filename)
        content.Split([|'\n'|], StringSplitOptions.RemoveEmptyEntries)
        |> Seq.fold f 0
    with
    | :? IOException as e ->
        Console.Error.WriteLine("Error reading file: " + e.Message)
        Environment.Exit(1)
        0

[<EntryPoint>]
let main args =
    if args.Length < 1 then
        usage()

    let filename = args.[0]
    let result = processFile filename
    Console.WriteLine("result = " + result.ToString());
    0 
